plugins/image-prioritizer/helper.php

```diff
@@ -136,6 +136,66 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
 	return $additional_properties;
 }
 
+/**
+ * Validates that the provided background image URL is valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param bool|WP_Error|mixed  $validity   Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+ * @return bool|WP_Error Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ */
+function image_prioritizer_filter_store_url_metric_validity( $validity, OD_Strict_URL_Metric $url_metric ) {
+	if ( ! is_bool( $validity ) && ! ( $validity instanceof WP_Error ) ) {
+		$validity = (bool) $validity;
+	}
+
+	$data = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+	if ( ! is_array( $data ) ) {
+		return $validity;
+	}
+
+	$r = wp_safe_remote_head(
+		$data['url'],
```

Comment:
Copied from https://github.com/WordPress/performance/pull/1697#discussion_r1859375043:

The `wp_safe_remote_head()` function passes the supplied URL through `wp_http_validate_url()`, but that doesn't necessarily mean the request is safe. I'm thinking there should perhaps be an allowlist of hosts that this is allowed to connect to. Maybe we query for the first image attachment and pass it into `wp_get_attachment_image_url()` to then extract the hostname from the image. This would be better than having to try to come up with a list of all image CDNs that are used by plugins.

The list of allowed hosts could also be sent to the frontend so as to avoid sending the invalid URL in the first place. See related todo:

https://github.com/WordPress/performance/blob/bc81588943e835707d70948bdd6a3c817c948a25/plugins/optimization-detective/detect.js#L361-L366

Then again, would be a shape to continuously reject all of the URL Metrics in the case that an unexpected host is being used for a given background image. In that case, perhaps we should allow `OD_URL_Metric` to be mutable, so in the case where an invalid host is provided this callback could instead do:

```php
$url_metric->set( 'lcpElementExternalBackgroundImage', null );
```

or rather

```php
$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```

When mutating an instance of `OD_URL_Metric` it should re-trigger the validation so that the mutation would cause the JSON Schema to be made invalid that it would throw an exception.

Also, whenever mutation happens then it should call a new `$this->group->clear_cache()` method which empties out `$group->result_cache` and which would in turn call `$collection->clear_cache()`.

So this `od_store_url_metric_validity` filter could be used in two ways:

1. A filter callback could return `false` or a non-empty `WP_Error` instance to block the URL Metric from being stored at all.
2. A filter callback could mutate the supplied `OD_URL_Metric` object to apply custom sanitizations or validation constraints which then would end up getting persisted in storage.